lib/SILOptimizer/Utils/Differentiation/Common.cpp

```diff
@@ -23,5 +23,178 @@ namespace autodiff {
 
 raw_ostream &getADDebugStream() { return llvm::dbgs() << "[AD] "; }
 
+//===----------------------------------------------------------------------===//
+// Code emission utilities
+//===----------------------------------------------------------------------===//
+
+void collectAllActualResultsInTypeOrder(
+    ApplyInst *ai, ArrayRef<SILValue> extractedDirectResults,
+    SmallVectorImpl<SILValue> &results) {
+  auto calleeConvs = ai->getSubstCalleeConv();
+  unsigned indResIdx = 0, dirResIdx = 0;
+  for (auto &resInfo : calleeConvs.getResults()) {
+    results.push_back(resInfo.isFormalDirect()
+                          ? extractedDirectResults[dirResIdx++]
+                          : ai->getIndirectSILResults()[indResIdx++]);
+  }
+}
+
+SILValue joinElements(ArrayRef<SILValue> elements, SILBuilder &builder,
```

Comment:
Is this really needed? Could we put this onto SILBuilder or something? Seems wrong to have such helper utilities.

Reply:
I would give it a name that makes it explicit that you are using a tuple. Otherwise, how do I know I am getting back a tuple?

Reply:
I renamed `joinElements` to `createTupleIfNotSingle` in https://github.com/apple/swift/pull/30970.  This helper is sufficiently ad-hoc that I wouldn't want to move it outside of differentiation to `SILBuilder`.

Feel free to give feedback on the PR!